pkg/client/patch.go

```diff
@@ -84,11 +84,15 @@ type MergeFromOptions struct {
 	OptimisticLock bool
 }
 
+type patchHandler interface {
+	createPatch(originalJSON, modifiedJSON []byte, dataStruct interface{}) ([]byte, error)
+}
+
 type mergeFromPatch struct {
-	patchType   types.PatchType
-	createPatch func(originalJSON, modifiedJSON []byte, dataStruct interface{}) ([]byte, error)
-	from        Object
-	opts        MergeFromOptions
+	patchType    types.PatchType
+	patchHandler patchHandler
```

Comment:
So the definition of "gomock friendly" is "can not use function types"?

Reply:
Generally speaking, everything that is unexported is considered to be an implementation detail consumers can not rely on. This is for example why cmp will refuse to compare private types, see https://github.com/google/go-cmp third bullet point

Reply:
Sure, I agree that it is an implementation details.
On the other side this PR tries to improve the user experience for users that consume the controller-runtime client pkg - in more details it improves the readability of unit tests with controller-runtime client that use gomock (the most popular mocking library in the golang ecosystem).

This is how an expectation for aPatch call with gomock looks like with the current master

```go
	expectedPatch := client.MergeFrom(mergeFrom)
	expectedData, expectedErr := expectedPatch.Data(obj)
	Expect(expectedErr).To(BeNil())

	c.EXPECT().Patch(ctx, obj, gomock.Any()).DoAndReturn(func(_ context.Context, _ client.Object, patch client.Patch, _ ...client.PatchOption) error {
		data, err := patch.Data(obj)
		Expect(err).To(BeNil())
		Expect(patch.Type()).To(Equal(expectedPatch.Type()))
		Expect(data).To(Equal(expectedData))
		return nil
	})
```

This is how it would look like with this PR:

```go
	 c.EXPECT().Patch(ctx, obj, client.MergeFrom(mergeFrom))
```

Reply:
How exactly do you assess in the second example that  the patch type is correct, that the data is correct and that it won't error if patch.Data is called?

Reply:
Why I have to asses this when the given client.Patch arg equals to the expected one? In my unit test I would be only interested if the client.Client is called with the expected args (expected obj, expected patch, etc). The fact that Patch func of client.Client internally is using the Data func of the passed patch arg is not something that I should verify in my unit test.

Reply:
Because you are looking at how the Patch looks, not at how it behaves even though you only care about the latter (and we export the latter and not the former). If we were to introduce a bug that made a patch always return no data, your tests would happily pass but your product would be broken.

Reply:
Yeah, but I think this is the fundamental principle of the unit testing. When you write a unit test, you are only interested in the functional correctness of your unit (your type). The functional correctness of the external types you use should be ensured with (unit) tests in the corresponding library/pkg.
That's why we have mocks and spies that help us to isolate the unit under test (our type) from all dependant types (external ones). When you start testing also the dependent types, this is no longer a unit test.

Reply:
Unit tests just means that you are testing one isolated unit (here: That Patch() is called with the correct value). You are testing for "correct value" by comparing two unexported struct instances. This is just invalid and has nothing to do with unit testing or not. The things you as a user of controller-runtime are allowed to make any assumptions about are exported.
Please feel free to bring this up in the next kubebuilder and controller-runtime meeting if you disagree.
/hold